Src/Newtonsoft.Json/JsonReader.cs

```diff
@@ -1109,5 +987,50 @@ public virtual void Close()
             _tokenType = JsonToken.None;
             _value = null;
         }
+
+        internal void ReadAndAssert()
+        {
+            if (!Read())
+            {
+                throw JsonSerializationException.Create(this, "Unexpected end.");
```

Comment:
On that basis, it may be as well to do so to the case JsonTextReader.ParseValue with the same message while I do so.